ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PointSchedule.hs

```diff
@@ -550,3 +551,17 @@ stToGen ::
 stToGen gen = do
   seed :: QCGen <- arbitrary
   pure (runSTGen_ seed gen)
+
+-- | Return the last block point served by the honest peer. Fails if the honest
+-- peer never sends a block, that is if they have no ticks or they only have
+-- 'NodeOffline' ticks.
+lastHonestBlockPoint :: PointSchedule -> WithOrigin TestBlock
+lastHonestBlockPoint = go Nothing . toList . ticks
+  where
+    go :: Maybe (WithOrigin TestBlock) -> [Tick] -> WithOrigin TestBlock
+    go Nothing [] = error "lastHonestBlockPoint"
+    go (Just bp) [] = bp
+    go _ (Tick{active=Peer HonestPeer (NodeOnline points)} : ticks) =
+      let AdvertisedPoints{block = BlockPoint bp} = points
+       in go (Just bp) ticks
+    go bp (_ : ticks) = go bp ticks
```

Comment:
An attempt without explicit recursion:
```Haskell
lastHonestBlockPoint s =
   let bps = mapMaybe honestBlockPoint $ toList (ticks s)
    in case reverse bps of
         [] -> error "lastHonestBlockPoint"
         bp:_ -> bp
  where
    honestBlockPoint Tick{active=Peer HonestPeer (NodeOnline points)} =
        let AdvertisedPoints{block = BlockPoint bp} = points
         in Just bp
    honestBlockPoint _ = Nothing
```

Reply:
With the last changes, this function has been made obsolete. I also ended up needing a more generic version in another piece of work which I think should be preferred to this `lastHonestBlockPoint`. See https://github.com/IntersectMBO/ouroboros-consensus/commit/61b84ef942d5357216ce828a0d078b4868ffffd1.